eventconsumer/src/main/scala/com/gu/notifications/events/AthenaLambda.scala

```diff
@@ -0,0 +1,125 @@
+      .build())
+    .build()
+}
+
+case class Query(database: String, queryString: String, outputLocation: String)
+
+class AthenaLambda {
+
+  import ExecutionContext.Implicits.global
+  private val envDependencies = new EnvDependencies
+  private val logger: Logger = LogManager.getLogger(classOf[AthenaLambda])
+  private val dynamoReportUpdater = new DynamoReportUpdater(envDependencies.stage)
+
+
+  def route(query: Query, startOfReportingWindow: ZonedDateTime): CompletableFuture[Void] = {
+    startQuery(query)
+      .thenComposeAsync(fetchQueryResponse)
+      .thenComposeAsync(updateDynamoIfRecent(_, startOfReportingWindow))
+      .thenAcceptAsync((aggregationCounts: AggregationCounts) => {
+        logger.info(s"Aggregation counts $aggregationCounts")
+        if (aggregationCounts.failure > 0) {
+          throw new RuntimeException("Failures")
+        }
+      })
+  }
+
+  def startQuery(query: Query): CompletableFuture[String] = {
+    val startQueryAndGetId: CompletableFuture[String] = executeQuery(query).thenApplyAsync((startQueryExecutionResponse: StartQueryExecutionResponse) => startQueryExecutionResponse.queryExecutionId())
+    startQueryAndGetId.thenComposeAsync((executionId: String) =>
+      waitUntilQueryCompletes(executionId).thenApplyAsync((_: Void) => executionId)
+    )
+  }
+
+
+  private def waitUntilQueryCompletes(id: String): CompletableFuture[Void] = {
+    Thread.sleep(1000)
+    athenaAsyncClient.getQueryExecution(GetQueryExecutionRequest.builder()
+      .queryExecutionId(id)
+      .build()).thenComposeAsync((response: GetQueryExecutionResponse) => {
+      response.queryExecution().status().state() match {
+        case QueryExecutionState.QUEUED => waitUntilQueryCompletes(id)
+        case QueryExecutionState.RUNNING => waitUntilQueryCompletes(id)
+        case _ => CompletableFuture.completedFuture(null)
+      }
+    })
+  }
+
+
+  private def updateDynamoIfRecent(notificationIdCounts: Map[String, PlatformCount], startOfReportingWindow: ZonedDateTime): CompletionStage[AggregationCounts] = {
+    val notificationReportEvents = notificationIdCounts.toList.map { case (id, count) => NotificationReportEvent(id, EventAggregation(count)) }
+    val resultCounts = dynamoReportUpdater.updateSetEventsReceivedAfter(notificationReportEvents, startOfReportingWindow)
+    val aggregatedCounts = AggregationCounts.aggregateResultCounts(resultCounts)
+    FutureConverters.toJava(aggregatedCounts)
+  }
+
+  private def fetchQueryResponse(id: String)(implicit executionContext: ExecutionContext): CompletableFuture[Map[String, PlatformCount]] = {
+    val queue = new ConcurrentLinkedQueue[List[(String, PlatformCount)]]()
+    athenaAsyncClient.getQueryResultsPaginator(GetQueryResultsRequest.builder()
+      .queryExecutionId(id)
+      .build()).subscribe((getQueryResultsResponse: GetQueryResultsResponse) => {
+      queue.add(getQueryResultsResponse.resultSet.rows().asScala.toList.tail.map { row =>
```

Comment:
`toList.tail` aren't you skipping a row here?

Reply:
Yes, the head is the headers; added a commit to print this as it makes the code more self-explanatory for future readers.